readthedocs/doc_builder/python_environments.py

```diff
@@ -72,6 +74,112 @@ def delete_existing_venv_dir(self):
+            # installation than ``docs`` user. Required for local building
+            # since everything is run as ``root`` when using Local Development
+            # instance
+            cmd = [
+                'ln',
+                '-s',
+                os.path.join(settings.RTD_DOCKER_WORKDIR, '.asdf'),
+                '/root/.asdf',
+            ]
+            self.build_env.run(
+                *cmd,
+            )
+
+        for tool, version in self.config.build.tools.items():
+            version = version.full_version  # e.g. 3.9 -> 3.9.7
+
+            # TODO: generate the correct path for the Python version
+            # tool_path = f'{self.config.build.os}/{tool}/2021-08-30/{version}.tar.gz'
+            tool_path = f'{self.config.build.os}-{tool}-{version}.tar.gz'
+            tool_version_cached = build_tools_storage.exists(tool_path)
+            if tool_version_cached:
+                remote_fd = build_tools_storage.open(tool_path, mode='rb')
+                with tarfile.open(fileobj=remote_fd) as tar:
+                    # Extract it on the shared path between host and Docker container
+                    extract_path = os.path.join(self.project.doc_path, 'tools')
+                    tar.extractall(extract_path)
+
+                    # Move the extracted content to the ``asdf`` installation
+                    cmd = [
+                        'mv',
+                        f'{extract_path}/{version}',
+                        os.path.join(
+                            settings.RTD_DOCKER_WORKDIR,
+                            f'.asdf/installs/{tool}/{version}',
+                        ),
+                    ]
+                    self.build_env.run(
+                        *cmd,
+                    )
+            else:
+                log.debug(
+                    'Cached version for tool not found. os=%s tool=%s version=% filename=%s',
+                    self.config.build.os,
+                    tool,
+                    version,
+                    tool_path,
+                )
+                # If the tool version selected is not available from the
+                # cache we compile it at build time
+                cmd = [
+                    # TODO: make this environment variable to work
+                    # 'PYTHON_CONFIGURE_OPTS="--enable-shared"',
+                    'asdf',
+                    'install',
+                    tool,
+                    version,
+                ]
+                self.build_env.run(
+                    *cmd,
+                )
+
+            # Make the tool version chosen by the user the default one
+            cmd = [
+                'asdf',
+                'global',
+                tool,
+                version,
+            ]
+            self.build_env.run(
+                *cmd,
+            )
+
+            # Recreate shims for this tool to make the new version
+            # installed available
```

Comment:
Could perhaps link to the asdf docs here? I don't know what `shims` are.

Reply:
`shims` is a standard concept that I don't really know how to explain 😓 but Wikipedia will help me here: https://en.wikipedia.org/wiki/Shim_(computing)

The documentation for asdf is at https://asdf-vm.com/learn-more/faq.html#newly-installed-exectable-not-running. I will put this in the comment as well.

Reply:
Actually, this `reshim` command may not be required in this case because it says that plugins automatically do this:

>  Those installed by plugins have shims automatically created